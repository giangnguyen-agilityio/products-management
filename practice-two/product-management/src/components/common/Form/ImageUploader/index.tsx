import React, { memo } from 'react'
import {
  FormControl,
  FormLabel,
  FormErrorMessage,
  Input,
  Image,
} from '@chakra-ui/react'
import { ERROR_MESSAGES } from '@constants/messages'

interface ImageUploaderProps {
  formData: {
    image: string
    name: string
  }
  handleImageUpload: (event: React.ChangeEvent<HTMLInputElement>) => void
}

const ImageUploader: React.FC<ImageUploaderProps> = ({
  formData,
  handleImageUpload,
}) => (
  <FormControl className="input-field image-uploaded" isRequired>
    <FormLabel
      className="form-label"
      fontFamily="Oswald-Regular"
      fontSize="xl"
      letterSpacing="1px"
    >
      Image:
    </FormLabel>
    <Input
      _hover={{ borderColor: 'primary' }}
      _focus={{ borderColor: 'primary' }}
      borderColor="gray.300"
      backgroundColor="background"
      padding={2}
      borderRadius="6px"
      height="45px"
      name="image"
      type="file"
      accept="image/*"
      onChange={handleImageUpload}
    />
    {formData.image && (
      <Image
        margin="10px 0"
        padding={4}
        border="1px solid"
        borderColor="primary"
        borderRadius={20}
        width="350px"
        objectFit="contain"
        className="image-uploaded"
        src={formData.image}
        alt={`The ${formData.name} image`}
      />
    )}
    <FormErrorMessage>
      {ERROR_MESSAGES.PRODUCT_IMAGE_IS_MISSING}
    </FormErrorMessage>
  </FormControl>
)

export default memo(ImageUploader)
