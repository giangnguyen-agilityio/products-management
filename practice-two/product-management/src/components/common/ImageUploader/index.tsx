import React, { memo } from 'react'
import { FormControl, FormLabel, Input, Image, Text } from '@chakra-ui/react'

// Define the prop types for the ImageUploader component
interface ImageUploaderProps {
  formData: {
    image: string
    name: string
  }
  handleImageUpload: (event: React.ChangeEvent<HTMLInputElement>) => void
  errorMessage: string
}

const ImageUploader: React.FC<ImageUploaderProps> = ({
  formData,
  handleImageUpload,
  errorMessage,
}) => (
  <FormControl className="input-field image-uploaded" isRequired>
    {/* Label for the image input */}
    <FormLabel
      className="form-label"
      fontFamily="Oswald-Regular"
      fontSize="xl"
      letterSpacing="1px"
    >
      Image:
    </FormLabel>
    {/* Input field for uploading an image */}
    <Input
      variant="primary"
      padding={2}
      height="45px"
      name="image"
      type="file"
      accept="image/*"
      onChange={handleImageUpload}
    />
    {/* Display the uploaded image */}
    {formData.image && (
      <Image
        margin="10px 0"
        padding={4}
        border="1px solid"
        borderColor="primary"
        borderRadius={20}
        width="350px"
        objectFit="contain"
        className="image-uploaded"
        src={formData.image}
        alt={`The ${formData.name} image`}
      />
    )}
    {/* Display an error message if the image is missing */}
    {/* Error message */}
    {errorMessage != null && <Text variant="danger">{errorMessage}</Text>}
  </FormControl>
)

export default memo(ImageUploader)
