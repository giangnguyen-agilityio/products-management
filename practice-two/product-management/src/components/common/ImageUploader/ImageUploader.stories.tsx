import type { Meta, StoryObj } from '@storybook/react'
import ImageUploader from './index'

const meta = {
  title: 'Practice Two/ImageUploader Component',
  component: ImageUploader,
  decorators: [
    Story => (
      <div
        style={{
          padding: '10px',
        }}
      >
        <Story />
      </div>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    handleImageUpload: { action: 'This image was uploaded' },
  },
} satisfies Meta<typeof ImageUploader>
export default meta
type Story = StoryObj<typeof meta>
export const FieldWithoutErrorMessage: Story = {
  args: {
    formData: {
      image: 'https://assets.nicepagecdn.com/d2cc3eaa/3159880/images/yttyy.jpg',
      name: 'The image alt description',
    },
    errorMessage: '',
  },
}

export const FieldWithErrorMessage: Story = {
  args: {
    formData: {
      image: '',
      name: '',
    },
    errorMessage: 'Please select an image',
  },
}
