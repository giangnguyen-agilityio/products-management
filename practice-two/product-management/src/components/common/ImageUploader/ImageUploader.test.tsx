import { render, fireEvent } from '@testing-library/react'
import ImageUploader from './'

describe('ImageUploader component', () => {
  const formData = {
    image: 'test-image-url',
    name: 'Test Image',
  }

  it('renders without errors', () => {
    const { container, getByText, getByLabelText, getByAltText } = render(
      <ImageUploader
        formData={formData}
        handleImageUpload={jest.fn()}
        errorMessage=""
      />
    )

    const imageInputLabel = getByText('Image:')
    const imageInput = getByLabelText('Upload Image')
    const uploadedImage = getByAltText('The Test Image image')

    expect(imageInputLabel).toBeInTheDocument()
    expect(imageInput).toBeInTheDocument()
    expect(uploadedImage).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('displays an error message', () => {
    const { getByText } = render(
      <ImageUploader
        formData={formData}
        handleImageUpload={jest.fn()}
        errorMessage="Image is required"
      />
    )

    const errorMessage = getByText('Image is required')
    expect(errorMessage).toBeInTheDocument()
  })

  it('calls handleImageUpload when an image is selected', () => {
    const handleImageUploadMock = jest.fn()
    const { getByLabelText } = render(
      <ImageUploader
        formData={formData}
        handleImageUpload={handleImageUploadMock}
        errorMessage=""
      />
    )

    const imageInput = getByLabelText('Upload Image')
    fireEvent.change(imageInput, {
      target: {
        files: [
          new File(['image content'], 'test-image.png', { type: 'image/png' }),
        ],
      },
    })

    expect(handleImageUploadMock).toHaveBeenCalled()
  })
})
