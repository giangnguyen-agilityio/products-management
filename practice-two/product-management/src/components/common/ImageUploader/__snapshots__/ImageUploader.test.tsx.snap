// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageUploader component renders without errors 1`] = `
<div>
  <div
    class="chakra-form-control input-field image-uploaded css-0"
    role="group"
  >
    <label
      class="chakra-form__label form-label css-16hu8s6"
      for="field-:r0:"
      id="field-:r0:-label"
    >
      Image:
      <span
        aria-hidden="true"
        class="chakra-form__required-indicator css-0"
        role="presentation"
      >
        *
      </span>
    </label>
    <input
      accept="image/*"
      aria-label="Upload Image"
      aria-required="true"
      class="chakra-input css-16crra2"
      id="field-:r0:"
      name="image"
      required=""
      type="file"
    />
    <img
      alt="The Test Image image"
      class="image-uploaded css-754yvm"
      src="test-image-url"
    />
    <p
      class="chakra-text css-0"
    />
  </div>
</div>
`;
